diesel_derives/src/sql_type.rs

```diff
@@ -0,0 +1,133 @@
+use quote::Tokens;
+use syn;
+
+use util::*;
+
+pub fn derive(item: syn::DeriveInput) -> Tokens {
+    let struct_ty = struct_ty(item.ident.clone(), &item.generics);
+    let item_name = item.ident.as_ref().to_uppercase();
+    let generics = &item.generics;
+    let pg_tokens = pg_tokens(&item.attrs, generics, &struct_ty);
+    let sqlite_tokens = sqlite_tokens(&item.attrs, generics, &struct_ty);
+    let mysql_tokens = mysql_tokens(&item.attrs, generics, &struct_ty);
+
+    wrap_item_in_const(
+        format!("_IMPL_SQL_TYPE_FOR_{}", item_name).into(),
+        quote!(
+            impl #generics diesel::types::NotNull for #struct_ty {}
+            impl #generics diesel::types::SingleValue for #struct_ty {}
+            #pg_tokens
+            #sqlite_tokens
+            #mysql_tokens
+        ),
+    )
+}
+
+fn pg_tokens(
+    attrs: &[syn::Attribute],
+    generics: &syn::Generics,
+    struct_ty: &syn::Ty,
+) -> Option<Tokens> {
+    use syn::{Lit, MetaItem, NestedMetaItem};
+
+    fn error() -> ! {
+        panic!(
```

Comment:
Didn't we decide [here](https://github.com/diesel-rs/diesel/pull/1327#pullrequestreview-80271119) that `compile_error!` messages are more readable?

Reply:
Very marginally, but it would be significantly harder to do that here than it was in that case